docs/design/phases.md

```diff
@@ -0,0 +1,110 @@
+
+We've decided to make the Kubernetes directory `/etc/kubernetes` a constant in the application, since it is clearly the given path in a lot of places,
+and the most intuitive location. Having that path configurable would for sure confuse readers of a on-top-of-kubeadm-implemented deployment solution.
+
+Further, having the admin/kubelet kubeconfig path would 
+
+### Certificates
+
+First off, kubeadm generates certificate and private key pairs for different purposes.
+Certificates are stored by default in `/etc/kubernetes/pki`.
+
+There should be:
+ - a CA certificate (`ca.crt`) with its private key (`ca.key`)
+ - an API Server certificate (`apiserver.crt`) using `ca.crt` as the CA with its private key (`apiserver.key`). The certificate should:
+   - be a serving server certificate (`x509.ExtKeyUsageServerAuth`)
+   - contain altnames for
+     - the kubernetes' service's internal clusterIP and dns name (e.g. `10.96.0.1`, `kubernetes.default.svc.cluster.local`, `kubernetes.default.svc`, `kubernetes.default`, `kubernetes`)
+     - the hostname of the node
+     - the IP address of the default gateway
+     - optional extra altnames that can be specified by the user
+   - be in the `system:masters` organization
+ - a client certificate for the apiservers to connect to the kubelets securely (`apiserver-kubelet-client.crt`) using `ca.crt` as the CA with its private key (`apiserver-kubelet-client.key`). The certificate should:
+   - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+   - be in the `system:masters` organization
+ - a certificate for signing ServiceAccount Tokens (`sa.crt`) using `ca.crt` as the CA with its private key (`sa.key`). The certificate should:
+   - ??
+
+If a given certificate and private key pair both exist, the generation will be skipped, and those files will validated and used if they are valid for the use-case.
+This means the user can for example give his own CA which then will be used signing the rest of the certs.
+
+### Generate KubeConfig files on the master
+
+There should be:
+ - a KubeConfig file for kubeadm to use itself and the admin: `/etc/kubernetes/admin.conf`
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - be in the `system:masters` organization
+ - a KubeConfig file for kubelet to use: `/etc/kubernetes/kubelet.conf`
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - be in the `system:nodes` organization
+
+The `ca.crt` is also embedded in the KubeConfig files.
+
+### Generate a token.csv file for bootstrapping purposes
+
+There should be:
+ - a `/etc/kubernetes/tokens.csv` file that contains one kubeadm-generated token
```

Comment:
@v1k0d3n the apiserver doesn't support reading tokens from json, so that's impossible for the time being

What we all want though, and what we're gonna implement is a dynamic way of managing these tokens so we can ditch this file totally (no one likes handling files on disk)